research/confidence_calibration/temperature_scaling.py

```diff
@@ -0,0 +1,379 @@
+            filename = Path(image_path).name
+            if filename in insect_labels.keys():
+                label_rank = insect_labels[filename]["taxon_rank"]
+                label_region = insect_labels[filename]["region"]
+                label_key = insect_labels[filename]["speciesKey"]
+
+                # Consider only species level labels
+                if (
+                    label_key
+                    and label_rank == "SPECIES"
+                    and label_region == region
+                    and label_key != -1
+                    and str(label_key) in category_map.keys()
+                ):
+                    with Image.open(image_path) as img:
+                        # Transform the image
+                        image = apply_transform_to_image(img)
+
+                        # Model inference
+                        logits = model(image.unsqueeze(0).to(device))
+
+                        # Save label in numerical format
+                        label = category_map[str(label_key)]
+
+                        # Append items
+                        labels_list.append(label)
+                        logits_list.append(logits)
+
+    # Concatenate all the logits and labels
+    labels = torch.Tensor(labels_list).cpu()
+    logits = torch.cat(logits_list).cpu()
+    # labels = torch.Tensor(labels_list)
+    # logits = torch.cat(logits_list)
+
+    del logits_list
+    del labels_list
+    gc.collect()
+
+    # Calculate ECE before and after calibration
+    ece_before_calibration = expected_calibration_error(
+        logits,
+        labels,
+        plot_reliability_diagram=plot_reliability_diagram,
+        reliability_diagram_title=reliability_diagram_title + "_before_calib_ami_traps",
+    )
+    ece_after_calibration = expected_calibration_error(
+        logits / optimal_t,
+        labels,
+        plot_reliability_diagram=plot_reliability_diagram,
+        reliability_diagram_title=reliability_diagram_title + "_after_calib_ami_traps",
+    )
+
+    return ece_before_calibration, ece_after_calibration
+
+
+def tune_temperature(
+    model: torch.nn.Module,
+    val_dataloader: torch.utils.data.DataLoader,
+    device: str,
+    initial_t: float = 1.5,
+    max_optim_iterations=50,
+    learning_rate=0.01,
+) -> float:
+    """Main temperature tuning function"""
+
+    model.eval()
+    logits_list = []
+    labels_list = []
+    with torch.no_grad():
+        for images, labels in val_dataloader:
+            images, labels = images.to(device), labels.to(device)
+            logits = model(images)
+            logits_list.append(logits)
+            labels_list.append(labels)
+
+    # Concatenate all the logits and labels
+    logits = torch.cat(logits_list).to(device)
+    labels = torch.cat(labels_list).to(device)
+    del logits_list
+    del labels_list
+    gc.collect()
+
+    temp_scaling = TemperatureScaling(initial_t=initial_t).to(device)
+    loss_function = nn.CrossEntropyLoss().to(device)
```

Comment:
Alright, so you are not setting the `reduction` arguments, which is set to `mean` by default.

Unless there is some counter-indication (which could be possible, I don't have enough knowledge about this), would you then be able to use `sum` instead, iterate on smaller portions of your dataset (or even item by item) and then calculate the mean manually at the end?